SDL2 Pthread Proxy APIs

jakogut opened the pull request:
This PR implements various C APIs for SDL2 to work with `PROXY_TO_PTHREAD=1` as discussed here: https://github.com/emscripten-ports/SDL2/pull/77

kripken commented:
Thanks @jakogut!

In general this looks great. Please also

 * Document these in `./site/source/docs/api_reference/emscripten.h.rst`
 * Add testing. Maybe a good place is to add it in `tests/test_html5.c`, and it looks like that file is tested both on the main thread and when proxied, so that should be enough.

jakogut commented:
@kripken I've added some documentation and tests for these APIs. These functions generally have void return types, so checking for success doesn't make a lot of sense. Let me know if these suffice, or if if there's a better way to test.

I also separated the proxied variant of the `sdl2_gl_read` test into a separate branch that will be submitted after the SDL2 fixes are merged.

kripken commented on src/library_browser.js:
```diff
@@ -1367,6 +1367,25 @@ var LibraryBrowser = {
+  emscripten_get_window_title__proxy: 'sync',
+  emscripten_get_window_title__sig: 'iv',
+  emscripten_get_window_title: function() {
+    return allocate(intArrayFromString(document.title), 'i8', ALLOC_STACK);
```
Allocating this on the stack is too risky, I think - if the pointer is used after the calling function unwinds the stack, it might be overwritten. How about having mallocing the space, and saving a single such allocation? (reusing it on the next call). And the docs can mention that the string is valid until the next call to the function.

jakogut replied:
Sounds good, is there an example of this I could use as a reference? It also looks like `_malloc` might be a better option here.

kripken replied:
I think this is a somewhat similar example: https://github.com/emscripten-core/emscripten/blob/7ba0b851d68544dee49c69231a260e26653bc233/src/library.js#L3808 , of a singleton allocation in a function.

kripken commented on src/library_browser.js:
```diff
@@ -1381,6 +1400,13 @@ var LibraryBrowser = {
+  emscripten_set_element_style: function(selector, property, value) {
+    var element = Module[selector];
+    if (element) element.style[property] = UTF8ToString(value);
+  },
```
I'm unclear on the benefit of this function. It looks on things on `Module` - what elements are expected to be there? Is it just the canvas?

kripken replied:
Also, if it's just the canvas, we already have `emscripten_set_canvas_element_size` which looks related?

jakogut replied:
> Also, if it's just the canvas, we already have `emscripten_set_canvas_element_size` which looks related?

Yeah, it could have a better name, perhaps `emscripten_set_canvas_element_style`? The purpose is to allow SDL2 to set the cursor style in a way that works with `PROXY_TO_PTHREAD`. You can see the intended usage here: https://github.com/emscripten-ports/SDL2/pull/77/commits/f0426362f751aabbda4f3d5e561e445566db4005.

kripken replied:
Thanks for the link! Now I see the issue here, I think.

For something like this, perhaps we can use `MAIN_THREAD_EM_ASM`? I see that isn't documented well, I'll look into that. But basically it lets you proxy a call to the main thread and run some EM_ASM code there.

jakogut replied:
That will work, I can remove this API, and update the SDL2 branch to use this syntax.

sbc100 commented:
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.

jakogut commented:
@kripken I've updated the `emscripten_get_screen_size` API to use a dynamically allocated buffer, and updated the documentation to reflect the change. I've also updated the tests to check the return values of the added functions, rather than just printing a message. Lastly, I've removed the `emscripten_set_element_style` API entirely.

kripken commented on src/library_browser.js:
```diff
@@ -1381,6 +1381,35 @@ var LibraryBrowser = {
   },
 
+  emscripten_get_window_title__proxy: 'sync',
+  emscripten_get_window_title__sig: 'iv',
```
this should be `vi` I think, not `iv`?

jakogut replied:
Doesn't the function signature indicate that it's currently correct?
`char *emscripten_get_window_title();`

However, this comment made me realize that the function signature in the documentation is incorrect, so I'll fix that.

kripken replied:
Oh, yes, maybe that's what confused me.

kripken commented on src/library_browser.js:
```diff
@@ -1381,6 +1381,35 @@ var LibraryBrowser = {
+  emscripten_get_window_title: function() {
+    var buflen = 256;
+
+    if (!_emscripten_get_window_title.buffer)
```
please use `{`, `}` as the body is on another line.

kripken commented on src/library_browser.js:
```diff
@@ -1381,6 +1381,35 @@ var LibraryBrowser = {
+      _emscripten_get_window_title.buffer
+    );
+
+    return _emscripten_get_window_title.buffer
```
missing `;` at the end.

kripken commented on src/library_browser.js:
```diff
@@ -1381,6 +1381,35 @@ var LibraryBrowser = {
+  emscripten_set_window_title__proxy: 'sync',
+  emscripten_set_window_title__sig: 'vi',
+  emscripten_set_window_title: function(title) {
+    setWindowTitle(UTF8ToString(title));
```
it seems a little odd to use `UTF8` here but `Ascii` while writing? Ascii for both is probably fine.

kripken commented on site/source/docs/api_reference/emscripten.h.rst:
```diff
@@ -493,6 +493,20 @@ Functions
 
+.. c:function:: char *emscripten_get_window_title()
+
+  Returns the window title
```
`.` at end of sentence please (also in places below)

kripken commented on src/library_browser.js:
```diff
@@ -1381,6 +1381,36 @@ var LibraryBrowser = {
+  emscripten_get_screen_size__sig: 'vii',
+  emscripten_get_screen_size: function(width, height) {
+    if (width) {{{ makeSetValue('width', '0', 'screen.width', 'i32') }}};
+    if (height) {{{ makeSetValue('height', '0', 'screen.height', 'i32') }}};
```
is it worth making these optional? we'd need to document that if we want it. but it seems simpler to assume the caller wants to receive both of them?

jakogut replied:
I think I'd also tend to agree that it's simpler to assume the caller wants both the width and height. This would be used specifically in the SDL2 port, in `Emscripten_VideoInit`[1], which gets the width and height in a single call. If the caller only wants one or the other, they can simply pass NULL for either parameter, and that value will be ignored.

https://github.com/jakogut/SDL2/blob/2b641edaad57e32a709a70a1609749ac26b6ec0c/src/video/emscripten/SDL_emscriptenvideo.c#L146

kripken replied:
Sorry for the delay - I missed this email somehow.

Sounds like this will be used where both width and height are used, so let's not make it optional? It just increases code size, requires new docs, and overall doesn't seem worth it.

I think the PR can land, once we fix that and there's a small typo comment from earlier! Thanks for the work here.

jakogut commented:
@kripken This is working in combination with modifications to [SDL2](https://github.com/emscripten-ports/SDL2/pull/127) as well as the below patch, a workaround for https://github.com/emscripten-core/emscripten/issues/7684. 

```patch
diff --git a/tests/sdl2_gl_frames_swap.c b/tests/sdl2_gl_frames_swap.c
index aff19e881..25e6ed4ed 100644
--- a/tests/sdl2_gl_frames_swap.c
+++ b/tests/sdl2_gl_frames_swap.c
@@ -7,6 +7,7 @@
 
 #include <SDL.h>
 #include <emscripten.h>
+#include <emscripten/html5.h>
 
 SDL_Window* window;
 SDL_Renderer* renderer;
@@ -30,6 +31,11 @@ void frame()
 
 int main()
 {
+    EmscriptenWebGLContextAttributes attr;
+    emscripten_webgl_init_context_attributes(&attr);
+    EMSCRIPTEN_WEBGL_CONTEXT_HANDLE ctx = emscripten_webgl_create_context("#canvas", &attr);
+    emscripten_webgl_make_context_current(ctx);
+
     SDL_Init(SDL_INIT_VIDEO);
     window = SDL_CreateWindow("test", 0, 0, 640, 480, SDL_WINDOW_OPENGL);
     renderer = SDL_CreateRenderer(window, 0, SDL_RENDERER_ACCELERATED);
```

Tested with:
`emcc tests/sdl2_gl_frames_swap.c -o sdl2_gl_frames_swap.html -s USE_SDL=2 -s USE_PTHREADS=1 -s OFFSCREEN_FRAMEBUFFER=1`

I also have a modified test case that I can update after this PR lands here https://github.com/emscripten-core/emscripten/pull/8282.

warvstar commented:
I've found that when adding -s PROXY_TO_PTHREAD=1 it gives the following error

```
Uncaught TypeError: Cannot read property 'getParameter' of undefined
    at emscriptenWebGLGet (o.js:8091)
    at _emscripten_glGetIntegerv (o.js:8292)
    at _do_call (<anonymous>:wasm-function[1628]:0x9418a)
    at emscripten_current_thread_process_queued_calls (<anonymous>:wasm-function[1627]:0x93c9a)
    at emscripten_main_thread_process_queued_calls (<anonymous>:wasm-function[1645]:0x94d7b)
    at o.js:1555
    at Worker.worker.onmessage (o.js:2037)
```

adding -s OFFSCREENCANVAS_SUPPORT=1 it then gives this error

```
Failed to execute 'getContext' on 'HTMLCanvasElement': Cannot get context from a canvas that has transferred its control to offscreen.
    at Object.createContext (http://localhost:3000/o.js:5764:20)
    at _eglCreateContext (http://localhost:3000/o.js:5947:24)
    at _emscripten_receive_on_main_thread_js (http://localhost:3000/o.js:9467:19)
    at _do_call (<anonymous>:wasm-function[1348]:0x8c66c)
    at emscripten_current_thread_process_queued_calls (<anonymous>:wasm-function[1347]:0x8bd8a)
    at emscripten_main_thread_process_queued_calls (<anonymous>:wasm-function[1365]:0x8ce5b)
    at http://localhost:3000/o.js:1555:22
    at Worker.worker.onmessage (http://localhost:3000/o.js:2045:13)
```

I was able to fix the second one by replacing the eglCreateContext call and making the existing context current via emscripten_webgl_make_context_current(handle); in SDL_render_gles2.c line ~1982.

However doing so gives me this error for OFFSCREEN_FRAMEBUFFER

```
Uncaught TypeError: Cannot set property '0' of undefined
    at Object.newRenderingFrameStarted (o.js:5714)
    at Browser_mainLoop_runner (o.js:4858)
```


I'd like to have support for OFFSCREEN_FRAMEBUFFER as well, anyone know the best place for me to start to make this work?
